### PR TITLE
[WIP] libSyntax: Use Parser to generate integer literal syntax.

### PR DESCRIPTION
**Pull request by nkcsgexi**
This patch tries to teach parser to generate libSytnax tree along with our existing ASTs. Adopting a bottom-up fashion, we generate syntax entity for integer literal expression as the start point.

### CODE REVIEW

File: include/swift/Syntax/SyntaxParsingContext.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+struct RawTokenInfo {
+  SourceLoc Loc;
+  RC<RawTokenSyntax> Token;
+};
+
+// The context that disables syntax tree building.
+class SyntaxParsingContext {
+public:
+  virtual void addTokenSyntax(SourceLoc Loc) {}
+  virtual void makeIntegerLiteralExp() {};
```
Comment (harlanhaskins) [inline include/swift/Syntax/SyntaxParsingContext.h]: Nit: `makeIntegerLiteralExpr`

**Comment by harlanhaskins** [comment]
@nkcsgexi Excited to see the infrastructure for this!

**Comment by nkcsgexi** [comment]
@harlanhaskins still several pending things before review : )

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@DougGregor @harlanhaskins Could you please take a look at this patch? does the infrastructure of parsing libSyntax entities make sense to you?

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: include/swift/AST/Module.h [hunk 2]
```diff
@@ -1088,9 +1090,22 @@ class SourceFile final : public FileUnit {
     return (bool)AllCorrectedTokens;
   }
 
+  syntax::SourceFileSyntax getSyntaxRoot() const {
+    return *SyntaxRoot;
```
Comment (harlanhaskins) [inline include/swift/AST/Module.h]: This would be worth an assertion, I think.

File: lib/Syntax/SyntaxParsingContext.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        getUnknownExpr({ *S.getAs<TokenSyntax>() }), None));
+    }
+    AllTopLevel.push_back(SyntaxFactory::makeTopLevelCodeDecl(
+      SyntaxFactory::makeStmtList(AllStmts)));
+  }
+
+  Trivia Leading, Trailing;
+  GlobalData.File.SyntaxRoot.emplace(
+    SyntaxFactory::makeSourceFile(SyntaxFactory::makeDeclList(AllTopLevel),
+      SyntaxFactory::makeToken(tok::eof, "\n", SourcePresence::Present,
```
Comment (harlanhaskins) [inline lib/Syntax/SyntaxParsingContext.cpp]: Nit: I think it'd be better here if we made the EOF empty and give it newline trivia. The text for EOF tokens isn't serialized.

```
Trivia Leading = Trivia::newlines(1), Trailing;
GlobalData.File.SyntaxRoot.emplace(
  SyntaxFactory::makeSourceFile(SyntaxFactory::makeDeclList(AllTopLevel),
    SyntaxFactory::makeToken(tok::eof, "\n", SourcePresence::Present, Leading, Trailing)));
```

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by gparker42** [comment]
I think this is causing build failures of the form "'swift/Syntax/SyntaxNodes.h' file not found". I'm going to revert it unless you have an immediate fix.
https://ci.swift.org/job/oss-swift_tools-RA_stdlib-DA_test-simulator/77/

**Comment by harlanhaskins** [comment]
@gparker42 I can fix this by forward-declaring `syntax::SourceFileSyntax` in `Module.h` which will remove the implicit dependency.

**Comment by harlanhaskins** [comment]
@gparker42 Ah, never mind, I can't forward declare this. The fix for this is more complex. Go ahead and revert and I'll find a good way to fix this.

**Comment by harlanhaskins** [comment]
The issue here is that the inclusion of `SyntaxNodes.h` in `Module.h` means that the downstream targets that include `Module.h` need to depend on `swift-syntax-generated-headers`. Unfortunately, _tons_ of downstream targets include `Module.h`.

**Comment by harlanhaskins** [comment]
So either we add that dependency to everyone that includes `Module.h`, or we..._sigh_...just commit the generated Syntax node files...

**Comment by nkcsgexi** [comment]
@gparker42 Thank you for reverting it! I'll take a look later.

**Pull request merged**